### PR TITLE
Resolve ChannelSpec into a Channel

### PR DESCRIPTION
**Pull request by AntoinePrv**
- [x] Based on #2870

### CODE REVIEW

File: libmamba/include/mamba/specs/authentication_info.hpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    /** A Conda token set in the URL path. */
+    struct CondaToken
+    {
+        std::string token;
+    };
+
+    using AuthenticationInfo = std::variant<BasicHTTPAuthentication, BearerToken, CondaToken>;
+
+    /**
+     * A class that holds the authetication info stotred by users.
```
Comment (JohanMabille) [inline libmamba/include/mamba/specs/authentication_info.hpp:40]:
```suggestion
     * A class that holds the authentication info stored by users.
```

File: libmamba/src/core/channel.cpp [hunk 2]
```diff
@@ -242,51 +112,29 @@ namespace mamba
     util::flat_set<std::pair<std::string, std::string>>
     Channel::platform_urls(bool with_credential) const
     {
-        std::string base = location();
-        if (with_credential && token())
+        if (!url().package().empty())
         {
-            base = util::join_url(base, "t", *token());
+            return {};
```
Comment (JohanMabille) [inline libmamba/src/core/channel.cpp:117]: Is an empty set expected in that case?

Comment (AntoinePrv) [reply libmamba/src/core/channel.cpp:117]: Yes. At least that is what I am aiming for. When specifiying a full URL to a package `https://mamba.org/somthing/pkg.conda` I am not assuming any repo/channel structure.

In the future, I'd like to better separate these two cases.

File: libmamba/include/mamba/util/tuple_hash.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2019, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_UTIL_TUPLE_HASH_HPP
+#define MAMBA_UTIL_TUPLE_HASH_HPP
+
```
[the remark this answers is not in the document]
Comment (AntoinePrv) [reply libmamba/include/mamba/util/tuple_hash.hpp:9]: I guess this discussion applies to a lot of the `util` library. We currently do not have a good way to both separate public/private headers **and** organize it by topic (util/specs/...).

**Review by JohanMabille — APPROVED** [review]

**Pull request merged**